### PR TITLE
Iter9

### PR DESCRIPTION
**Pull request by shekshuev**

### CODE REVIEW

File: cmd/shortener/main.go [hunk 1]
```diff
@@ -6,16 +6,22 @@ import (
 
 	"github.com/shekshuev/shortener/internal/app/config"
 	"github.com/shekshuev/shortener/internal/app/handler"
+	"github.com/shekshuev/shortener/internal/app/logger"
+	"go.uber.org/zap"
```
Comment (arttet) [inline cmd/shortener/main.go]: [CHANGE REQUEST] Please sort imports

File: internal/app/compress/gzip.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+func (c *GzipWriter) Header() http.Header {
+	return c.w.Header()
+}
+
+func (c *GzipWriter) Write(p []byte) (int, error) {
+	return c.zw.Write(p)
+}
+
+func (c *GzipWriter) WriteHeader(statusCode int) {
+	if statusCode < 300 {
```
Comment (arttet) [inline internal/app/compress/gzip.go]: [CHANGE REQUEST] This is a magic number

File: internal/app/store/url_store.go [hunk 3]
```diff
@@ -34,3 +74,29 @@ func (s *URLStore) GetURL(key string) (string, error) {
 	}
 	return value, nil
 }
+
+func (s *URLStore) persist(key string) error {
```
Comment (arttet) [inline internal/app/store/url_store.go]: [CHANGE REQUEST] Please follow the Single-responsibility principle

File: internal/app/store/url_store.go [hunk 4]
```diff
@@ -1,19 +1,59 @@
...
 
 var ErrEmptyKey = fmt.Errorf("key cannot be empty")
 var ErrEmptyValue = fmt.Errorf("value cannot be empty")
 var ErrNotFound = fmt.Errorf("not found")
 
-func NewURLStore() *URLStore {
-	return &URLStore{urls: make(map[string]string)}
+func NewURLStore(cfg *config.Config) *URLStore {
+	store := &URLStore{urls: make(map[string]string), cfg: cfg}
+	if _, err := os.Stat(cfg.FileStoragePath); err == nil {
```
Comment (arttet) [inline internal/app/store/url_store.go]: [CHANGE REQUEST] Please follow the Single-responsibility principle

File: internal/app/middleware/gzip.go [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+import (
+	"net/http"
+	"strings"
+
+	"github.com/shekshuev/shortener/internal/app/compress"
+)
+
+func GzipCompressor(h http.HandlerFunc) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		ow := w
```
Comment (arttet) [inline internal/app/middleware/gzip.go]: [CHANGE REQUEST] This name is too short.

File: internal/app/logger/logger.go [hunk 6]
```diff
@@ -0,0 +1,22 @@
+package logger
+
+import (
+	"go.uber.org/zap"
+)
+
+var Log *zap.Logger = zap.NewNop()
```
Comment (arttet) [inline internal/app/logger/logger.go]: [CHANGE REQUEST] Please don't use global variables

File: internal/app/handler/url_handler.go [hunk 7]
```diff
@@ -13,11 +16,21 @@ type URLHandler struct {
 	Router  *chi.Mux
 }
 
+type Middleware func(http.HandlerFunc) http.HandlerFunc
+
+func conveyor(h http.HandlerFunc, middlewares ...Middleware) http.HandlerFunc {
```
Comment (arttet) [inline internal/app/handler/url_handler.go]: [CHANGE REQUEST] Let's simplify it.

File: internal/app/handler/url_handler.go [hunk 8]
```diff
@@ -43,6 +56,38 @@ func (h *URLHandler) createURLHandler(w http.ResponseWriter, r *http.Request) {
 	}
 }
 
+func (h *URLHandler) createURLHandlerJSON(w http.ResponseWriter, r *http.Request) {
+	if r.Method == http.MethodPost {
```
Comment (arttet) [inline internal/app/handler/url_handler.go]: [CHANGE REQUEST] Please follow the Go style

File: internal/app/store/url_store.go [hunk 9]
```diff
@@ -1,19 +1,59 @@
...
+	"github.com/shekshuev/shortener/internal/app/config"
+	"github.com/shekshuev/shortener/internal/app/logger"
 )
 
 type URLStore struct {
 	urls map[string]string
+	cfg  *config.Config
+}
+
+type serializeData struct {
```
Comment (arttet) [inline internal/app/store/url_store.go]: [CHANGE REQUEST] It should be in the `dto` package.

**Review by arttet — CHANGES_REQUESTED** [review]

File: internal/app/logger/logger.go [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+	"go.uber.org/zap"
+)
+
+type Logger struct {
+	Log *zap.Logger
+}
+
+var (
+	instance *Logger
```
Comment (arttet) [inline internal/app/logger/logger.go]: [CHANGE REQUEST] Please don't use global variables

File: internal/app/store/url_store_snapshot.go [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+)
+
+func CreateSnapshot(store *URLStore) error {
+	file, err := os.OpenFile(store.cfg.FileStoragePath, os.O_CREATE|os.O_WRONLY, 0644)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	for key, value := range store.urls {
```
Comment (arttet) [inline internal/app/store/url_store_snapshot.go]: [CHANGE REQUEST] Please follow the Open–closed principle

File: internal/app/store/url_store.go [hunk 12]
```diff
@@ -24,7 +36,7 @@ func (s *URLStore) SetURL(key, value string) error {
 		return ErrEmptyValue
 	}
 	s.urls[key] = value
-	return nil
+	return CreateSnapshot(s)
```
Comment (arttet) [inline internal/app/store/url_store.go]: [CHANGE REQUEST] We should do it only once at the end of the program's execution.

**Review by arttet — CHANGES_REQUESTED** [review]

File: internal/app/handler/url_handler.go [hunk 13]
```diff
@@ -15,41 +18,64 @@ type URLHandler struct {
...
 }
 
 func (h *URLHandler) getURLHandler(w http.ResponseWriter, r *http.Request) {
-	if r.Method == http.MethodGet {
-		if longURL, err := h.service.GetLongURL(r.URL.Path[1:]); err == nil {
-			http.Redirect(w, r, longURL, http.StatusTemporaryRedirect)
-		} else {
-			w.WriteHeader(http.StatusBadRequest)
-		}
+	if longURL, err := h.service.GetLongURL(r.URL.Path[1:]); err == nil {
```
Comment (arttet) [inline internal/app/handler/url_handler.go]: `r.URL.Path[1:]` it looks like unsafe.

File: internal/app/logger/logger.go (hunk 10 above)
Comment (arttet) [reply internal/app/logger/logger.go]: This comment remains relevant.

File: internal/app/store/url_store_snapshot.go [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+	"bufio"
+	"encoding/json"
+	"os"
+
+	"github.com/google/uuid"
+	"github.com/shekshuev/shortener/internal/app/models"
+)
+
+func (s *URLStore) CreateSnapshot() error {
+	file, err := os.OpenFile(s.cfg.FileStoragePath, os.O_CREATE|os.O_WRONLY, 0644)
```
Comment (arttet) [inline internal/app/store/url_store_snapshot.go:13]: [SUGGESTION] This solution will be difficult to maintain and extend, violates SOLID principles, and distorts the essence of the task.

**Review by arttet — CHANGES_REQUESTED** [review]

File: internal/app/logger/logger.go (hunk 10 above)
Comment (shekshuev) [reply internal/app/logger/logger.go]: паттерн синглтон, поэтому глобальная переменная

File: internal/app/store/url_store_snapshot.go (hunk 14 above)
Comment (shekshuev) [reply internal/app/store/url_store_snapshot.go:13]: утверждение субъективно и не обосновано. Хотелось бы более подробного ответа, почему так

**Review by Bazys — COMMENTED** [review]
у меня будет только один комментарий - папка app лишняя в го.

**Pull request merged**